### PR TITLE
Enemy firing system refactor

### PR DESCRIPTION
**Pull request by TokyoSU**
Sorry for the branch name, but there was too much change about gunflash, i will move the cowboy from it and create another branch just for him.

- Fixed rendering of gunflash.
- Now entity can shoot with 2 weapon (left and right) like lara.

Should be what @Lwmte proposed in #532

### CODE REVIEW

File: TombEngine/Game/Lara/lara_fire.cpp [hunk 1]
```diff
@@ -786,10 +786,10 @@ FireWeaponType FireWeapon(LaraWeaponType weaponType, ItemInfo& targetEntity, Ite
 	auto& player = *GetLaraInfo(&laraItem);
 	auto& ammo = GetAmmo(player, weaponType);
 
-	if (ammo.GetCount() == 0 && !ammo.HasInfinite())
+	if (ammo.GetCount() == 0 && !ammo.HasInfinite() && weaponType != LaraWeaponType::Snowmobile)
```
Comment (Lwmte) [inline TombEngine/Game/Lara/lara_fire.cpp:789]: I think that maybe better solution will be assigning ammo to snowmobile as infinite? This looks like a hack.

File: TombEngine/Game/Lara/lara_fire.cpp [hunk 2]
```diff
@@ -786,10 +786,10 @@ FireWeaponType FireWeapon(LaraWeaponType weaponType, ItemInfo& targetEntity, Ite
...
 		return FireWeaponType::NoAmmo;
 
-	if (!ammo.HasInfinite())
+	if (!ammo.HasInfinite() && weaponType != LaraWeaponType::Snowmobile)
```
Comment (Lwmte) [inline TombEngine/Game/Lara/lara_fire.cpp:792]: As above.

File: TombEngine/Game/control/box.cpp [hunk 3]
```diff
@@ -722,12 +726,32 @@ short CreatureTurn(ItemInfo* item, short maxTurn)
 	return angle;
 }
 
+static void PlayGunEffectForCreature(ItemInfo* item, const CreatureMuzzleflashInfo& muzzleFlash)
+{
+	if (muzzleFlash.Delay != 0)
```
Comment (Lwmte) [inline TombEngine/Game/control/box.cpp:731]: Do an early exit there by inverting condition.

File: TombEngine/Game/control/box.cpp [hunk 4]
```diff
@@ -722,12 +726,32 @@ short CreatureTurn(ItemInfo* item, short maxTurn)
 	return angle;
 }
 
+static void PlayGunEffectForCreature(ItemInfo* item, const CreatureMuzzleflashInfo& muzzleFlash)
```
Comment (Lwmte) [inline TombEngine/Game/control/box.cpp:729]: Why `static`?

File: TombEngine/Game/control/box.h [hunk 5]
```diff
@@ -99,10 +50,10 @@ struct OVERLAP
 	int flags;
 };
 
-struct BiteInfo
+/*struct BiteInfo
```
Comment (Lwmte) [inline TombEngine/Game/control/box.h:53]: Don't leave commented blocks like that. If this code will be needed in future, add a comment tagged with `TODO` in front. Otherwise, delete it.

File: TombEngine/Game/control/lot.cpp [hunk 6]
```diff
@@ -146,6 +146,12 @@ void InitialiseSlot(short itemNumber, bool makeTarget)
 
 			break;
 
+		case LotType::SnowmobileGun:
```
Comment (Lwmte) [inline TombEngine/Game/control/lot.cpp:149]: What this new LOT type has to do with enemy firing system refactor? If you have introduced snowmobile gun enemy, it should be on a separate PR.

File: TombEngine/Game/effects/effects.cpp [hunk 7]
```diff
@@ -52,7 +52,7 @@ int SplashCount = 0;
 Vector3i NodeVectors[ParticleNodeOffsetIDs::NodeMax];
 NODEOFFSET_INFO NodeOffsets[ParticleNodeOffsetIDs::NodeMax] =
 {
-	{ -16, 40, 160, -LM_LHAND, false }, // TR5 offset 0, TODO: This mesh is invalid as it can't be negative. -- TokyoSU 23.02.20
+	{ -16, 40, 160, 13, false },		// TR5 offset 0
```
Comment (Lwmte) [inline TombEngine/Game/effects/effects.cpp:55]: Is this bone index for Lara? If it is, it should use enum, otherwise it should not and everything is okay until we demagic enemy bone indices.

File: TombEngine/Game/effects/smoke.cpp [hunk 8]
```diff
@@ -191,6 +191,51 @@ namespace TEN::Effects::Smoke
 		s.room = LaraItem->RoomNumber;
 	}
 
+	void TriggerGunSmokeParticles(ItemInfo* item, int x, int y, int z, int xv, int yv, int zv, byte initial, byte count, short roomNumber)
```
Comment (Lwmte) [inline TombEngine/Game/effects/smoke.cpp:194]: Can we make another overload of `TriggerGunSmokeParticles` utilize this variation under the hood to avoid copypaste?

File: TombEngine/Game/itemdata/creature_info.h [hunk 9]
```diff
@@ -71,6 +71,60 @@ struct LOTInfo
 	bool CanMonkey	  = false;
 };
 
+struct CreatureBiteInfo
+{
+	Vector3i Position = Vector3i::Zero;
+	int BoneID = -1;
```
Comment (Lwmte) [inline TombEngine/Game/itemdata/creature_info.h:77]: We have `NO_JOINT` constant which can be utilized here instead of magic number.

File: TombEngine/Game/people.cpp [hunk 10]
```diff
@@ -81,7 +81,7 @@ short GunMiss(int x, int y, int z, short velocity, short yRot, short roomNumber)
 {
 	auto pos = GameVector(
 		LaraItem->Pose.Position.x + ((GetRandomControl() - 0x4000) << 9) / 0x7FFF,
-		LaraItem->Floor,
+		LaraItem->Floor - 128,
```
Comment (Lwmte) [inline TombEngine/Game/people.cpp:84]: Why -128 is introduced here?

File: TombEngine/Objects/TR2/Entity/tr2_knife_thrower.cpp [hunk 11]
```diff
@@ -264,7 +233,7 @@ namespace TEN::Entities::Creatures::TR2
 
 				if (!creature->Flags)
 				{
-					CreatureEffect(item, KnifeBiteLeft, ThrowKnife);
+					CreatureEffect2(item, KnifeBiteLeft, 100, torso, ThrowKnife);
```
Comment (Lwmte) [inline TombEngine/Objects/TR2/Entity/tr2_knife_thrower.cpp:236]: What's the difference between `CreatureEffect` and `CreatureEffect2`? If that's not an original Core naming, I'd introduce a bit more descriptive name for this function.

File: TombEngine/Objects/TR2/Entity/tr2_worker_dualrevolver.cpp [hunk 12]
```diff
@@ -223,9 +228,11 @@ namespace TEN::Entities::Creatures::TR2
 					extraTorsoRot.y = AI.angle;
 				}
 
-				if (!creature->Flags)
+				if (creature->Flags == 0 && item->Animation.FrameNumber == GetFrameIndex(item, 0))
```
Comment (Lwmte) [inline TombEngine/Objects/TR2/Entity/tr2_worker_dualrevolver.cpp:231]: We have `TestAnimFrame` function for this purpose.

File: TombEngine/Objects/TR2/Vehicles/skidoo.cpp [hunk 13]
```diff
@@ -663,21 +666,20 @@ namespace TEN::Entities::Vehicles
...
-			if ((int)FireWeapon(LaraWeaponType::Pistol, *lara->TargetEntity, *laraItem, angles) +
-				(int)FireWeapon(LaraWeaponType::Pistol, *lara->TargetEntity, *laraItem, angles))
-			{
-				skidoo->FlashTimer = 2;
-				SoundEffect(weapon.SampleNum, &laraItem->Pose);
-				skidooItem->ItemFlags[0] = 4;
-			}
+			FireWeapon(LaraWeaponType::Snowmobile, *lara->TargetEntity, *laraItem, angles);
+			FireWeapon(LaraWeaponType::Snowmobile, *lara->TargetEntity, *laraItem, angles);
+			//lara->LeftArm.GunFlash = 1;
```
Comment (Lwmte) [inline TombEngine/Objects/TR2/Vehicles/skidoo.cpp:679]: What are these two commented lines doing?

File: TombEngine/Specific/setup.cpp [hunk 14]
```diff
@@ -45,254 +44,6 @@ void InitialiseGameFlags()
 	Camera.underwater = false;
 }
 
-void ObjectObjects()
```
Comment (Lwmte) [inline TombEngine/Specific/setup.cpp:48]: Why this subfunction was merged into parent one? This decreases readability, and also it is not related to this PR purpose.

File: TombEngine/Specific/setup.cpp [hunk 15]
```diff
@@ -45,254 +44,6 @@ void InitialiseGameFlags()
...
-		if (obj->loaded)
-		{
-			obj->initialise = InitialiseShootSwitch;
-			obj->control = ControlAnimatingSlots;
-			obj->collision = ShootSwitchCollision;
-		}
-	}
-}
-
-void TrapObjects()
```
Comment (Lwmte) [inline TombEngine/Specific/setup.cpp:230]: Where this function was moved to? I don't see this code reintroduced somewhere else.

**Review by Lwmte — CHANGES_REQUESTED** [review]

File: TombEngine/Specific/setup.cpp (hunk 15 above)
Comment (TokyoSU) [reply TombEngine/Specific/setup.cpp:230]: To tr5_objects.cpp

File: TombEngine/Specific/setup.cpp (hunk 14 above)
Comment (TokyoSU) [reply TombEngine/Specific/setup.cpp:48]: Was leftover when i was doing the objects folder with every objects files.

File: TombEngine/Objects/TR2/Vehicles/skidoo.cpp (hunk 13 above)
Comment (TokyoSU) [reply TombEngine/Objects/TR2/Vehicles/skidoo.cpp:679]: It was a test for the snowmobile gunflash, but it's hardcoded, the flash was positioned at lara hand, not the snowmobile so i commented it for later.

File: TombEngine/Objects/TR2/Entity/tr2_worker_dualrevolver.cpp (hunk 12 above)
Comment (TokyoSU) [reply TombEngine/Objects/TR2/Entity/tr2_worker_dualrevolver.cpp:231]: oh sorry will change that

File: TombEngine/Objects/TR2/Entity/tr2_knife_thrower.cpp (hunk 11 above)
Comment (TokyoSU) [reply TombEngine/Objects/TR2/Entity/tr2_knife_thrower.cpp:236]: CreatureEffect use item->Animation.Velocity.z for speed, CreatureEffect2 let the programmer specify it in the argument.

File: TombEngine/Game/control/lot.cpp (hunk 6 above)
Comment (TokyoSU) [reply TombEngine/Game/control/lot.cpp:149]: It's to fix snowmobile driver, actually it could only go forward and drop/climb 1 click, in TR2, it was able to fall from more than 4 click but not climb 2 click.
it's a small think so i through to let it on this PR. unlike cowboy which was a new entity.

File: TombEngine/Game/effects/effects.cpp (hunk 7 above)
Comment (TokyoSU) [reply TombEngine/Game/effects/effects.cpp:55]: Node offset is not for lara, i think it was for a entity and was just a typo.

File: TombEngine/Game/people.cpp (hunk 10 above)
Comment (TokyoSU) [reply TombEngine/Game/people.cpp:84]: the bullet which miss the target was inside the floor not on it so i added that.

File: TombEngine/Objects/Generic/Traps/traps.cpp [hunk 16]
```diff
@@ -1,299 +0,0 @@
-#include "framework.h"
```
[the remark this answers is not in the document]
Comment (TokyoSU) [reply TombEngine/Objects/Generic/Traps/traps.cpp:1]: traps was not used anywhere so i just removed it, the function inside was already somewhere else.

File: TombEngine/Game/effects/smoke.cpp (hunk 8 above)
Comment (TokyoSU) [reply TombEngine/Game/effects/smoke.cpp:194]: Wanted to have something fast to test, i think @Sezzary will be able to change it correctly and @Nemoel-Tomo to upgrade it to look nice ingame.

File: TombEngine/Game/control/box.cpp (hunk 3 above)
Comment (TokyoSU) [reply TombEngine/Game/control/box.cpp:731]: Done.

File: TombEngine/Game/control/box.cpp (hunk 4 above)
Comment (TokyoSU) [reply TombEngine/Game/control/box.cpp:729]: It's not used elsewhere and i think it won't be.

File: TombEngine/Game/control/box.h (hunk 5 above)
Comment (TokyoSU) [reply TombEngine/Game/control/box.h:53]: Done.

File: TombEngine/Game/effects/smoke.cpp (hunk 8 above)
Comment (TokyoSU) [reply TombEngine/Game/effects/smoke.cpp:194]: Done, just added roomNumber as a new argument.

File: TombEngine/Game/itemdata/creature_info.h (hunk 9 above)
Comment (TokyoSU) [reply TombEngine/Game/itemdata/creature_info.h:77]: Done.

File: TombEngine/Objects/TR2/Entity/tr2_worker_dualrevolver.cpp (hunk 12 above)
Comment (TokyoSU) [reply TombEngine/Objects/TR2/Entity/tr2_worker_dualrevolver.cpp:231]: Done.

File: TombEngine/Game/itemdata/creature_info.h [hunk 17]
```diff
@@ -71,6 +71,60 @@ struct LOTInfo
 	bool CanMonkey	  = false;
 };
 
+struct CreatureBiteInfo
```
Comment (Sezzary) [inline TombEngine/Game/itemdata/creature_info.h:74]: Since we're renaming this, why don't we come up with a name better than "bite"? :P

**Comment by Stranger1992** [comment]
I cannot seem to build this ?

I am getting ```Severity	Code	Description	Project	File	Line	Suppression State
Error (active)	E0135	class "EntityModelData" has no member "Mutator"	TombEngine	C:\GitHub\TombEngine\TombEngine\Objects\TR2\Entity\tr2_dragon.cpp	667	
```

as an error

File: TombEngine/Game/control/box.cpp [hunk 18]
```diff
@@ -32,6 +35,7 @@ constexpr auto FEELER_DISTANCE = CLICK(2);
 constexpr auto FEELER_ANGLE = ANGLE(45.0f);
 constexpr auto CREATURE_AI_ROTATION_MAX = ANGLE(90.0f);
 constexpr auto CREATURE_JOINT_ROTATION_MAX = ANGLE(70.0f);
+constexpr auto GUN_EFFECT_CREATURE_YSHIFT = 75;
```
[the remark this answers is not in the document]
Comment (TokyoSU) [reply TombEngine/Game/control/box.cpp:38]: Used in PlayGunEffectForCreature() :)

**Review by Kubsy — APPROVED** [review]
